### DISCUSSION TITLE
RFC: BcNuGet package format

### DISCUSSION THREAD

**Discussion by freddydk** [microsoft/AL-Go#301] [Ideas]
## BcNuGet
We want to create a standardized way of exchanging Business Central apps across different DevOps solutions. How can partners expose their apps or runtime packages as NuGet packages, which easily can be consumed by other partners.
Presentation from last weeks meeting: [How do we solve the issue of having dependencies to AppSource apps.pdf](https://github.com/microsoft/AL-Go/files/10055847/How.do.we.solve.the.issue.of.having.dependencies.to.AppSource.apps.pdf)

The description should also allow Microsoft to host a NuGet Server, in which we could host all AppSource Apps if we are allowed to do so by partners.

This discussion is an attempt to define the BcNuGet package format and also describe how the implementation of this should work in DevOps solutions.

## Assumptions
Every BcNuGet will contain one app and a list of dependencies to other BcNuGet packages (apps) from that app.
A BcNuGet package can contain the app or runtime packages
If a BcNuGet package contains runtime packages, it must contain all necessary runtime packages (for all compatible versions) - see more under runtime packages
Test Apps can also be published as seperate BcNuGet packages if partners want you to run their test apps in your DevOps pipelines
 
## File Format
```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd">
    <metadata>
        <id>$id</id>
        <version>$version</version>
        <title>$brief</title>
        <description>$description</description>
        <authors>$publisher</authors>
        <repository type="git" url="$repository" branch="$branch" commit="$sha" />
        <licenseUrl>$EULA</licenseUrl>
        <projectUrl>$projectUrl</projectUrl>
        <icon>$icon</icon>
        <readme>$readme</readme>
        <releaseNotes>$releaseNotes</releaseNotes>
        <businessCentral privacyStatement="$privacyStatement" help="$help" url="$url" runtime="runtime" platform="$platform" application="$application" ... >
            <idRanges>
                <idRange from="$from" to="$to" />
            </idRanges>
            <resourceExposurePolicy allowDebugging="$allowDebugging" allowDownloadingSource="$allowDownloadingSource" ...  />
        </businessCentral>
        <dependencies>
            <dependency id="$id" version="$version" />
        </dependencies>
    </metadata>
    <files>
        <file src="$publisher_$name_$version.app" target="$publisher_$name_$version.app" />
        <file src="$platform\$publisher_$name_$version.runtime.app" target="$platform\$publisher_$name_$version.runtime.app" />
        <file src="$icon" target="$icon" />
        <file src="$readme" target="$readme" />
    </files>
</package>
```

In general, fields that are optional in app.json are also optional in the BcNuGet spec (like icon). Description however is a requirement in nuspec and if the app.json file doesn't contain a description, we will use the name instead.

The BusinessCentral tag is the only "custom" nuspec field, which then contains all Business Central specific properties.
Attributes on the Business Central tag are the simple app.json properties (mandatory properties are mandatory)
resourceExposurePolicy tag contains a copy of the resourceExposurePolicy tag from app.json as xml
idRanges tag contains a copy of the idRanges from app.json as xml

**Where**
$id, $version, $brief, $description, $publisher and $EULA are the corresponding values from app.json
$rpid is the runtime package identifier for that specific runtime package
$icon is the path and name of the icon within the package (optional)
$readme is the path and name of the readme.md file within the package(optional)
$projectUrl is a URL to the project, which hosts this project
$repository, $branch, $sha is the git repository, branch and sha from where the package was created
$releaseNotes markdown content of release notes for this version (max. 35000 characters)
$from, $to are object range from..to fields from app.json

## Platform
- Describes which platform the runtime package inside the folder was created for
- Microsoft will provide information/code, which can identify which different platforms we need to create runtime packages for
- Microsoft will provide information/code, to identify which runtime package to use for a given platform version

## Dependency handling
Default behavior of NuGet dependency resolution can be found here: https://learn.microsoft.com/en-us/nuget/concepts/dependency-resolution
Business Central doesn't really have a dependency resolution mechanism, but works after the rather simplified assumption that if you set a dependency to version 2.3.0.0 - then you are compatible with 2.3.0.0 and all subsequent versions (also 3.0.0.0 and 4.0.0.0 etc)
So, what is the right approach?

Let's have a look at this scenario?
![image](https://user-images.githubusercontent.com/10775043/203016976-47a9d897-ffed-40d1-b5b2-16fe45bcddd2.png)
What version of App B and App C do you get if you install App A, 1.0.0.0?
My suggestion would be to install App B version 1.0.1.0 - which is the latest build of version 1.0, which you have a dependency on.
Even if you are running BC version 20, as App B version 1.0.1.0 also should be compatible with 20.x

What version of App B and App C do you get if you install App A, 2.0.0.0?
My suggestion would be to install App B and App C version 2.0.0.0. When trying to install on 19.0, it would fail before starting the installation. If you already have App C 1.0.0.0 installed in the environment, it is upgraded to 2.0.0.0

What version of App B and App C do you get if you install App A, 2.1.0.0?
My suggestion would be to install App B version 2.1.0.0 and App C version 2.0.0.0. When trying to install on 19.0, it would fail before starting the installation. If you already have App C 1.0.0.0 installed in the environment, it is upgraded to 2.0.0.0

**Comment by azwierzchowskitnp** [comment]
## Introduction

In my opinion, we should make sure that NuGet works not only with DevOps but also covers other cases. I would prefer situation where NuGet server is the only source of apps inside a company. It means that people should also be able to go there, find and download packages. I understand that we are talking about DevOps here, but preparing for other scenarios will make this solution better and allow more people to use it.

We want to use DevOps to create NuGet packages and then allow developers/consultants to access to that server to search and download apps. We are also going to use our own version of "Extension Management" page to install and update apps directly from NuGet servers with automatic dependency resolution inside WebClient. It will make creation of demo/dev/test environments easier and help us with customers servers maintenance (if there is no DevOps "installation" task).

I was also showing an idea of NuGet version of "Download Symbols" command on my Directions session which could help developers to quickly switch between latest and lowest version of dependencies to check if their solution compiles. That idea could also be used in DevOps to quickly run that check before we create a docker (fail fast approach).

## Let's make it compatible with the specification

I would prefer if we make sure that our packages are 100% compatible with NuGet format and all NuGet tools/servers work without any problems. Unfortunately, it is not the case with the idea of adding `businessCentral` node to the .nuspec file. When I tried to create package using NuGet CLI (nuget.exe), it failed with a message that `businessCentral` is an unexpected node in the file.

NuSpec specification can be found here:
https://learn.microsoft.com/en-us/nuget/reference/nuspec

Also, I don't see any benefit from keeping that information in .nuspec. Fortunately, it is ignored by AzureDevOps and GitHub NuGet repositories, so you can upload these incompatible packages there, but we should not assume that it will work with other hosting providers. NuGet servers have very nice set of APIs that allows to search and see packages properties (i.e., we can build list of dependencies to download without downloading any package), but because `businessCentral` node is not part of the specification, it is not included in these APIs. The only solution at this moment is to use Tags to add additional properties to the packages.

### Proposed solution

#### Use NavxManifest.xml or app.json instead of `businessCentral` node

Instead of including `businessCentral` node in .nuspec file, we can include `NavxManifest.xml` from .app package or `app.json` in the NuGet package. 

``` xml
<file src="NavXManifest.xml" target="NavXManifest.xml" />
```

or

``` xml
<file src="app.json" target="app.json" />
```

#### Use tags for important BC App properties

We can use `<tags></tags>` node that contains space separated list of tags for additional, BC-specific attributes, that we would like to check before we download a package. Obviously we will have to agree what tags we would like to use there.

Here is an example of valid .nuspec file that uses these tags:
- bcAppId:`applicationId` - `id` property value form app.json
- bcAppPackageType:`packageType`  - `SingleApp` for one app in the package, but we can think about other scenarios like `MultipleApps`, `ConfigPackages`
- bcAppRuntime:`runtime` - `runtime` property value from app.json
- bcAppApplication:`application` - `application` property value from app.json
- bcAppPlatform:`platform` - `platform` property value from app.json

``` xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd" xmlns:bcns="http://schemas.microsoft.com/packaging/2022/11/bcspec.xsd">
  <metadata>
    <id>fad4be71-bf29-4ff6-e4e1-18d3968ff689</id>
    <version>1.10.0</version>
    <title>Customer extension</title>
    <authors>Partner</authors>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <description>Partner - Customer extension</description>
    <repository type="git" url="https://github.com/anzwdev/ALNuGetExperiments">
    </repository>
    <dependencies>
      <dependency id="bae47ec8-63aa-48f6-a4e1-90d3f68b66e1" version="1.0.0" />
      <dependency id="846de607-f93a-433d-940e-130a1830f422" version="1.0.0" />
    </dependencies>
    <tags>bcAppId:fad4be71-bf29-4ff6-e4e1-18d3968ff689 bcAppPackageType:SingleApp bcAppRuntime=8.0 bcAppPlatform:18.0.0.0 bcAppApplication:18.0.0.0</tags>
  </metadata>
  <files>
    <file src="Partner_Customer extension_1.10.0.0.app" target="Partner_Customer extension_1.10.0.0.app" />
    <file src="NavXManifest.xml" target="NavXManifest.xml" />
  </files>
</package>
```

And here is the result of a call to one of APIs on GitHub NuGet server. As you can see, we can find our tags there and we can use them to decide if we should download the package (i.e., by looking at the 'bcAppRuntime' or 'bcAppPlatform' tags)

``` json
{
    "@id": "https://nuget.pkg.github.com/AnZwDev/fad4be71-bf29-4ff6-e4e1-18d3968ff689/1.10.0.json",
    "packageContent": "https://nuget.pkg.github.com/AnZwDev/download/fad4be71-bf29-4ff6-e4e1-18d3968ff689/1.10.0/fad4be71-bf29-4ff6-e4e1-18d3968ff689.1.10.0.nupkg",
    "catalogEntry": {
        "@id": "https://nuget.pkg.github.com/AnZwDev/fad4be71-bf29-4ff6-e4e1-18d3968ff689/1.10.0.json",
        "authors": "Partner",
        "copyright": "",
        "dependencyGroups": [
            {
                "@id": "https://nuget.pkg.github.com/AnZwDev/fad4be71-bf29-4ff6-e4e1-18d3968ff689/1.10.0.json#dependencygroup",
                "@type": "PackageDependencyGroup",
                "dependencies": [
                    {
                        "@id": "https://nuget.pkg.github.com/AnZwDev/fad4be71-bf29-4ff6-e4e1-18d3968ff689/1.10.0.json#dependencygroup/bae47ec8-63aa-48f6-a4e1-90d3f68b66e1",
                        "@type": "PackageDependency",
                        "id": "bae47ec8-63aa-48f6-a4e1-90d3f68b66e1",
                        "range": "[1.0.0, )"
                    },
                    {
                        "@id": "https://nuget.pkg.github.com/AnZwDev/fad4be71-bf29-4ff6-e4e1-18d3968ff689/1.10.0.json#dependencygroup/846de607-f93a-433d-940e-130a1830f422",
                        "@type": "PackageDependency",
                        "id": "846de607-f93a-433d-940e-130a1830f422",
                        "range": "[1.0.0, )"
                    }
                ]
            }
        ],
        "description": "Partner - Customer extension",
        "iconUrl": "",
        "id": "fad4be71-bf29-4ff6-e4e1-18d3968ff689",
        "isPrerelease": false,
        "language": "",
        "licenseUrl": "",
        "packageContent": "https://nuget.pkg.github.com/AnZwDev/download/fad4be71-bf29-4ff6-e4e1-18d3968ff689/1.10.0/fad4be71-bf29-4ff6-e4e1-18d3968ff689.1.10.0.nupkg",
        "projectUrl": "",
        "requireLicenseAcceptance": false,
        "summary": "",
        "tags": "bcAppId:fad4be71-bf29-4ff6-e4e1-18d3968ff689 bcAppPackageType:SingleApp bcAppRuntime=8.0 bcAppPlatform:18.0.0.0 bcAppApplication:18.0.0.0",
        "version": "1.10.0"
    }
}
```

## Readability

If you take a look at .nuspec and api call results and think how NuGet package files are named, you will probably see that we might have a readability problem there. It won't be an issue for automatic DevOps scripts, but it will be a problem for people doing any manual work (investigation, search, downloads, manual installation). Package files names are created from package id and version number. If we use just BC App Id for the NuGet package ID, we will have a lot of GUIDs without any explanation what they mean. If you download a few packages, you will see GUIDs and versions on the list of your files, if you open .nuspec files, you will see a list of GUIDs on the list of dependencies without really knowing what these dependencies are.

Maybe we should use a bit more for the package id? What if we use AppId + Publisher + Name? Maybe we could remove spaces from Publisher and replace spaces with dots in the Name to make it look like any other, non-al package. But be aware that NuGet server might have a limit on the Id length - NuGet.org has 128 characters limit, so I would truncate the id to 128 characters.

If we do it this way, the .nuspec from example above could look a bit more readable:

``` xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd" xmlns:bcns="http://schemas.microsoft.com/packaging/2022/11/bcspec.xsd">
  <metadata>
    <id>fad4be71-bf29-4ff6-e4e1-18d3968ff689.Patner.Customer.extension</id>
    <version>1.10.0</version>
    <title>Customer extension</title>
    <authors>Partner</authors>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <description>Partner - Customer extension</description>
    <repository type="git" url="https://github.com/anzwdev/ALNuGetExperiments">
    </repository>
    <dependencies>
      <dependency id="bae47ec8-63aa-48f6-a4e1-90d3f68b66e1.ISV1.Warehouse.Integration" version="1.0.0" />
      <dependency id="846de607-f93a-433d-940e-130a1830f422.ISV2.Web.Integration" version="1.0.0" />
    </dependencies>
    <tags>bcAppId:fad4be71-bf29-4ff6-e4e1-18d3968ff689 bcAppPackageType:SingleApp bcAppRuntime=8.0 bcAppPlatform:18.0.0.0 bcAppApplication:18.0.0.0</tags>
  </metadata>
  <files>
    <file src="Partner_Customer extension_1.10.0.0.app" target="Partner_Customer extension_1.10.0.0.app" />
    <file src="NavXManifest.xml" target="NavXManifest.xml" />
  </files>
</package>
```

I know that it breaks the current BC Apps dependency rules where we can change publisher and name at any time but makes maintenance much easier.

## Pre-release versions

Maybe we coud use pre-release version suffix to mark some of our packages as pre-release? Nuget tools (i.e., nuget.exe) ignore these versions until you specify that you want to include them in search/installation. They are created by adding "-" and a text to the version number (i.e., 1.10.2.3-beta, 1.10.2.7-development, 2.6.3.5-customertest). This way we could reuse existing NuGet feeds to host these packages and define different "pre-release suffixes" in the pipelines. It will also help us with preventing a situation when a package is installed in an incorrect database.

**Reply by jonaswre** [reply]
> Let's make it compatible with the specification

I agree keeping in spec is better. This is the only way to guarantee ongoing support on all nuget servers - including Github and DevOps. 
But for resolution the requirements are to specialized to support nuget.exe.
For example resolution of packages that are compatible with a specific runtime will probably need some custom logic.

> I know that it breaks the current BC Apps dependency rules where we can change publisher and name at any time but makes maintenance much easier.

I think this should not be done. Adding a constraint that was removed by Microsoft feels wrong. 
Changing the name of an app could break the dependency resolution. The only identifier should be the app id. I think for manual search the title should be enough. From my experience with the Azure Feed implementation if the right tooling is in place nobody looks at the feed anyway. This is the beauty of automatic resolution. 

>bcAppPackageType:packageType - SingleApp for one app in the package, but we can think about other scenarios like MultipleApps, ConfigPackages

Could you explain what you mean by ConfigPackages? 

As for MultipleApps I think this should not be endorsed. 
The deployable unit within Business Central is an app.  Each app is installed individually. 
To group multiple apps would allow for multiple versions of an app to be relased at the same time. 
Imagaine you download two groups - if these groups contain some of the same apps which should you use. If the anwser is the latest whats the reason to group them in the first place? 
Grouping apps makes sense in development because changes in multiple apps can be developed together. 
In deployment this grouping is not needed when using nuget. Nuget automaticly gets all the apps that are needed without grouping them in delivery. 

AppSource has the concept of libary apps - But I never understood them because they have the same problem. Which libary is used if I provide the same for two apps but in diffrent versions?  Maybe somebody can explain that to me?

I think it would be better to also upload these libaries indiviually in AppSource and mark them as libary - but thats a diffrent topic.

**Reply by freddydk** [reply]
A few thoughts - after reading this the first time - will go through your emails more thoroughly.

Yes, lets stick to the spec (I actually thought it was extensible since this worked on GitHub).
Tags is a good idea for the important app.json properties, which we want to be able to search and filter on.

On the naming. When we get to the point where Microsoft can host all AppSource apps, I think Microsoft should use AppIds for naming.
When partners are hosting BcNuGet packages for their customers, I think they can name them differently.
This means that partners will give you a package name - and that package name will not change - and if your nuget package have a dependency on another nuget package - that will also have a name (and the server needs to be trusted)

Maybe the format should support partners in hosting packages with more than 1 .app file - but again, the Microsoft NuGet server will eventually be an extract of AppSource - and if you have a dependency on an AppId, then we can find it and traverse, but for private servers, multiple apps in one package could make sense for some partners.

and... for the Microsoft NuGet Server - the UI will be AppSource - partners will not be looking in a NuGet UI for apps - but for partner servers they might.

**Reply by jonaswre** [reply]
> When partners are hosting BcNuGet packages for their customers, I think they can name them differently. 

I think is this would defeat the purpose of a spec. I understood the spec to be able to interop between trusted partners. If a partner can name their packages differently or provide more then one app. Resolving dependencies between feeds can't be automated. 

Maybe the tooling can allow this if a partner desires but it should not be part of any spec.

**Reply by azwierzchowskitnp** [reply]
>But for resolution the requirements are to specialized to support nuget.exe.

Yes, we will have to implement our own logic here. 

>Could you explain what you mean by ConfigPackages?

The main subject of this discussion is using NuGet with DevOps for packaging single apps, but I wanted to mention different possible scenarios that can help us to promote NuGet usage in our community. If we have a NuGet server, we can always use it to deliver other  files to our databases. Imagine that you implement NuGet API client in AL - you can then use it to download and import data from your servers.

I would add the `bcAppPackageType` tag with `SingleApp` value only to the specification and a note that if somebody wants to use NuGet packages for any other scenarios, they should use their own value in that tag, so our dependency resolution logic can ignore these other, special packages.

> When partners are hosting BcNuGet packages for their customers, I think they can name them differently.

I agree with @jonaswre here - for a single app package we need one clear rule how we can convert dependency entry from app.json into dependency entry in .nuspec file. When somebody adds a new dependency to the app.json and pushes code to the git server, devops pipeline should be able to update .nuspec and find that dependency package without any manual action.

**Reply by freddydk** [reply]
IMO, it really isn't a matter of whether it is a single-app package or not, but more whether or not it is on Microsoft hosted NuGet Server.
If it is on the Microsoft hosted NuGet Server - then we will only have single app packages, named by ID's and dependencies are by ID's.
The search engine for the Microsoft hosted NuGet Server is AppSource - so no UI needed in that (IMO).

IMO, for partner hosted NuGet Servers - you can have multi or single app packages. A feed is given as any name (ex. BusinessCentralApps-BingMaps.AppSource) + a server url and a token (or it can be on public NuGet.org), but you specify that you trust that feed by taking a dependency on it (+ any dependencies it specifies)
The dependencies from this package can be other feed names, which must be on servers you trust, or it can be by AppId to the Microsoft server which you always trust.

So, for AppSource packages (if Microsoft hosts the server), everything will just work.
Until Microsoft hosts a server, partners will have to do this themselves. Even after Microsoft hosts a server, partners can opt-out and do this themselves.
For PTEs or apps from partner servers, consumers will need a server, a token and a feed name from the partner.
He needs to specify that he trusts the server+token+feed to download packages. If the feed contains dependencies to other feeds on the same server it is auto-trusted. If there is a dependency on a 4th party server - he needs to trust that as well.

I don't see why naming of self-hosted packages needs to be like the Microsoft hosted server (which I think is just AppId)?

**Reply by freddydk** [reply]
BTW, on this statement:
_I agree with @jonaswre here - for a single app package we need one clear rule how we can convert dependency entry from app.json into dependency entry in .nuspec file. When somebody adds a new dependency to the app.json and pushes code to the git server, devops pipeline should be able to update .nuspec and find that dependency package without any manual action._

I think it all depends what dependency you add.
In the world where Microsoft hosts a trusted server, and you add a dependency to an app on this server - then yes.
If you add a dependency to one of your own apps, then yes.
If you add a dependency to an app from a 3rd party vendor, which is included in any of the feeds you already specify, then yes.
If you add a dependency to a new app from a new vendor, I think you should specify which feed you depend on in your settings as well as the appid in app.json.

The reasons are many for this, but first and foremost security.
It is OK to search a trusted server for packages with a specific name (here I think Microsoft 
It is OK to get additional apps added in trusted feeds.
It is OK to get additional feeds added in trusted feeds, but it needs to resolve to the same package (at all partners) - 100% of the time - if we built-in some search algorithm here, we open an attack surface.
It is not OK to add additional feeds just by adding a dependency to app.json (unless they are on a trusted server), as this again opens up for attack surfaces,

As an example, you could imagine one evil developer in a large feature check-in is adding a dependency to App B from App A, but he modified one character in the AppID. Since name and publisher states the other app - everything looks good to the code reviewer, but suddenly customers using App A have another app included from a feed somewhere, which could lead to problems.

Yes, this should be caught by code reviews and people investigating what apps are included, but when we move towards automation; we need to trust the systems responsible for the automation.

Jonas came with a suggestion what we could check the signature on the app (certificate), but that would only help with appsource apps - and those will eventually get solved by the Microsoft hosted server anyway - so we still need to solve this for on-premises and other apps.

**Reply by jonaswre** [reply]
@freddydk I agree with you adding new feeds should not be automated. this is the only manual work besides adding the dependency to the app.json. But I think  for the spec its important how everything is structured. 

My dream situation is not even that everything works magically. I think it's not to much effort to require somebody to specify where they want to get their apps from. 
I think the most common use case is getting access to unreleased app versions before they hit the app source so that a partner can prepare for changes before they hit the app source. 

> Jonas came with a suggestion what we could check the signature on the app (certificate), but that would only help with appsource apps - and those will eventually get solved by the Microsoft hosted server anyway - so we still need to solve this for on-premises and other apps.

From a security standpoint we should verify every app we download it doesn't matter what app you download. The reason for this is that a partner could, even by accident publish an app of a another partner in their own feed, which might then be resolved beforehand.

I think to make it mandatory to sign every app you publish into a feed is not outrageous. A self signed certificate could also be used if we don't want the paywall of buying a certificate. 

Every partner can keep a list of which publisher has which trusted certificate signature. It then wouldn't matter where you get your apps from because you could verify that each app was created by the intended publisher. 

This is how Microsoft Updates must also work if you download them from untrusted system, for example your colleges computer. 

I think using this mechanism the only assumption I make is I need to trust the content of the partners app which I need to do in every case. But I don't need to trust the source where I got it from.

**Reply by freddydk** [reply]
> I think the most common use case is getting access to unreleased app versions before they hit the app source so that a partner can prepare for changes before they hit the app source.

Good and important scenario.
I do think that should eventually also be part of appsource (preview)

The problem with self-signed certificates is that they cannot be revoked. So, I trust a partner who loses his certificate - and now I potentially trust evil people,
And with trusted certificates, which are not self-signed needs to be refreshed every x years - meaning that all partners needs to update a key for the partner - right?

**Reply by jonaswre** [reply]
Self Signed Certificates can't be revoked that's true. But they could be excluded from the list of trusted certificates. I imagine the list as not trusted by default. 

My current thinking is - a partner is maintaining their own list which would indeed mean that every partner needs to update that list every x years. 

Solution for this is obviously a trusted registry for these certificate-publisher combinations. Maybe maintained by Microsoft as part of the AppSource? 

Another solution could be partner maintained list were changes need to be aknowledged by the user.

**Reply by freddydk** [reply]
IMO Certificates could definitely be a 2nd line of defense - but I don't think it can be the primary line of defense.
Any solution where partners will actually download potentially evil stuff, just to discover that it isn't trusted won't pass our security review (pretty sure about that).
But to be sure, I will investigate how Microsoft does this with other things.
PowerShell modules, NuGet packages, etc.

**Reply by jonaswre** [reply]
That's basically the concept of File hash verification. 
You download the exe and before execution you verify if what you have downloaded is actually what you wanted to download. 

I'm interested to see what Microsoft does in these cases there must be an acceptable  solution for this problem.
